misc/python/materialize/mzbuild.py

```diff
@@ -212,6 +212,63 @@ def inputs(self) -> set[str]:
         return set(git.expand_globs(self.rd.root, f"{self.source}/{self.matching}"))
 
 
+class PSUploadSources(PreImage):
+    """A `PreImage` action which uploads a source tarball to S3."""
+
+    def __init__(self, rd: RepositoryDetails, path: Path, config: dict[str, Any]):
+        super().__init__(rd, path)
+
+        bin = config.pop("bin", None)
+        if bin is None:
+            raise ValueError("mzbuild config is missing 'bin' argument")
+        self.bin_path = path / bin
+        self.out_path = path / "sources.tar.zstd"
+
+    def run(self) -> None:
+        super().run()
+        with open(self.bin_path, "rb") as bin:
+            build_id = get_build_id(bin)
+
+        cmd1 = ["/usr/bin/llvm-dwarfdump", "--show-sources", self.bin_path]
+        cmd2 = [
+            "/usr/bin/tar",
+            "-cf",
+            self.out_path,
+            "--zstd",
+            "-T",
+            "-",
+            "--ignore-failed-read",
+        ]
+
+        p1 = subprocess.Popen(cmd1, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        p2 = subprocess.Popen(
+            cmd2, stdin=p1.stdout, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        p1.stdout.close()
```

Comment:
Can you explain the intended stdout/stderr handling a bit more? In particular:

* What does closing the stdout stream do?
* Why are we `PIPE`ing both streams when we don't use them later?
* In case of an error, how do we get access to the error messages for debugging?

Reply:
to be completely honest, I created this from an example given by ChatGPT, because otherwise I had no idea how to make a pipeline in Python. I'll clean it up a bit and add comments.

Reply:
BTW, I already discussed this some [on Slack](https://materializeinc.slack.com/archives/C01G8JWQPB8/p1700071520485399) before putting up the PR, and it seems this is the morally correct route (though maybe you're right that we can exclude stderr piping).

As for the error message, the error handling here works the same way as in our `spawn.runv` library function, so I think it's OK.

Reply:
Yes, looks good now! (I think there is no need to `PIPE` the stdout of `p2`, but probably doesn't matter.)